### PR TITLE
feat: Cache piercing for discovery and async-dns

### PR DESCRIPTION
**Pull request by johanandren**
This makes it possible to programatically force re-resolution even if TTL is long, needed downstream in Akka gRPC for periodic refresh of nodes used for client side load balancing.

### CODE REVIEW

File: akka-actor/src/main/scala/akka/io/SimpleDnsCache.scala [hunk 1]
```diff
@@ -74,12 +74,18 @@ class SimpleDnsCache extends Dns with PeriodicCacheCleanup with NoSerializationV
   }
 
   @tailrec
-  private[io] final def put(key: (String, RequestType), records: Resolved, ttl: CachePolicy): Unit = {
+  private[akka] final def put(key: (String, RequestType), records: Resolved, ttl: CachePolicy): Unit = {
```
Comment (johanandren) [inline akka-actor/src/main/scala/akka/io/SimpleDnsCache.scala:77]: Opened up slightly to be able to write test in discovery

**Review by patriknw — APPROVED** [review]
LGTM, after minor

**Comment by johanandren** [comment]
Something I've thought about during the weekend: 

There is a race between TTL shutdown and the janitor actor cleaning up, we could potentially do better by having an I'm-going-to-stop callback to the Topic that it calls just before shutting down and then have a retry spin if the name is still occupied but not in the registry where the actor will soon shut down.

Not entirely sure it is worth it though.

**Comment by johanandren** [comment]
Ups, wrong issue!

**Comment by leviramsey** [comment]
I'm wondering if an alternative approach of setting a maximum DNS TTL would be conceptually cleaner than exposing a drop operation on the cache through service discovery?

**Comment by johanandren** [comment]
This is not intended as general purpose but a means for a specific downstream feature in Akka gRPC for discovering updated sets of services to load balance over. Forcing all cache to be dropped for that purpose seems wasteful and would also not be enough for that feature since we also need the drop to trigger a refreshed list to push into the gRPC client.

**Comment by johanandren** [comment]
For the record, this is the Akka gRPC functionality PR: https://github.com/akka/akka-grpc/pull/1886

**Pull request merged**